constraint/pkg/core/constraints/constraints_test.go

```diff
@@ -0,0 +1,123 @@
+package constraints
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func TestSemanticEqual(t *testing.T) {
+	testCases := []struct {
+		name string
+		c1   *unstructured.Unstructured
+		c2   *unstructured.Unstructured
+		want bool
+	}{
+		{
```

Comment:
nit: a testcase for passing the same pointer might also be useful, as that's covered by your logic